packages/controller-utils/src/siwe.ts

```diff
@@ -34,10 +34,130 @@ function msgHexToText(hex: string): string {
   }
 }
 
+/**
+ * @type WrappedSIWERequest
+ *
+ * Sign-In With Ethereum (SIWE)(EIP-4361) message with request metadata
+ * @property {string} from - Subject account address
+ * @property {string} origin - The RFC 3986 originating authority of the signing request, including scheme
+ * @property {ParsedMessage} siwe - The data parsed from the message
+ */
+export interface WrappedSIWERequest {
+  from: string;
+  origin: string;
+  siwe: SIWEMessage;
+}
+
+interface HostParts extends DomainParts {
+  scheme: string;
+}
+
+interface DomainParts {
+  userinfo: string | null;
+  host: string;
+  port: number | null;
+}
+
+export const ORIGIN_REGEX =
+  /^(?<scheme>[^:/]+):\/\/((?<userinfo>[^@]*)@)?(?<host>[^:@/?]+)(:(?<port>[0-9]+))?/u;
+export const DOMAIN_REGEX =
+  /^((?<userinfo>[^@]*)@)?(?<host>[^:@/?]+)(:(?<port>[0-9]+))?$/u;
+const DEFAULT_PORTS_BY_SCHEME = {
+  http: 80,
+  https: 443,
+} as { [key: string]: number | undefined };
+
+/**
+ * Parses parts from RFC 3986 authority from environment relevant for EIP-4361 origin validation.
+ *
+ * @param authority - input string
+ * @returns parsed parts
+ */
+export const parseAuthorityParts = (authority: string): HostParts => {
+  const match = authority.match(ORIGIN_REGEX);
+  if (!match) {
+    throw new Error(`Unrecognized origin format "${origin}".`);
+  }
+  const { scheme, userinfo, host, port } = match.groups as {
+    scheme: string;
+    userinfo: string | undefined;
+    host: string;
+    port: string | undefined;
+  };
+  return {
+    scheme,
+    // strip password from userinfo
+    userinfo: userinfo ? userinfo.replace(/:.*$/u, '') : null,
+    host,
+    port: port ? parseInt(port) : DEFAULT_PORTS_BY_SCHEME[scheme] || null,
+  };
+};
+
+/**
+ * Parses parts from RFC 3986 authority from EIP-4361 `domain` field.
+ *
+ * @param domain - input string
+ * @returns parsed parts
+ */
+export const parseDomainParts = (domain: string): DomainParts => {
+  const match = domain.match(DOMAIN_REGEX);
+  if (!match) {
+    throw new Error(`Unrecognized domain format "${domain}".`);
+  }
```

Comment:
I think there are some invalid domain formats getting past the regex. That said, if these domain outliers do parse into parts, they should not match the origin parts and the user should see a warning modal as expected. 

<img width="563" alt="Screen Shot 2023-04-12 at 12 03 16 AM" src="https://user-images.githubusercontent.com/20778143/231340991-e16cdf55-8c58-47a7-a9ef-7d225a85e4b8.png">

example invalid domain values that pass:
- 'this_is_an_invalid_domain' 
- '-example-domain-' 
- a domain name that exceeds the char limits